libImaging/SgiRleDecode.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * The Python Imaging Library.
+ * $Id$
+ *
+ * decoder for Sgi RLE data.
+ *
+ * history:
+ * 2017-07-28 mb    fixed for images larger than 64KB
+ * 2017-07-20 mb	created
+ *
+ * Copyright (c) Mickael Bonfill 2017.
+ *
+ * See the README file for information on usage and redistribution.
+ */
+
+#include "Imaging.h"
+#include "Sgi.h"
+
+#define SGI_HEADER_SIZE 512
+#define RLE_COPY_FLAG 0x80
+#define RLE_MAX_RUN 0x7f
+
+static void read4B(UINT32* dest, UINT8* buf)
+{
+    *dest = (UINT32)((buf[0] << 24) | (buf[1] << 16) | (buf[2] << 8) | buf[3]);
+}
+
+static int expandrow(UINT8* dest, UINT8* src, int n, int z)
+{
+    UINT8 pixel, count;
+    
+    for (;n > 0; n--)
+    {
+        pixel = *src++;
+        if (n == 1 && pixel != 0)
+            return n;
+        count = pixel & RLE_MAX_RUN;
+        if (!count)
+            return count;
+        if (pixel & RLE_COPY_FLAG) {
+            while(count--) {
+                *dest = *src++;
+                dest += z;
+            }
+                
+        }
+        else {
+            pixel = *src++;
+            while (count--) {
+                *dest = pixel;
+                dest += z;
+            }
+        }
+        
+    }
+    return 0;
+}
+
+static int expandrow2(UINT16* dest, UINT16* src, int n, int z)
+{
+    UINT8 pixel, count;
+
+   
+    for (;n > 0; n--)
+    {
+        pixel = ((UINT8*)src)[1];
+        ++src;
+        if (n == 1 && pixel != 0)
+            return n;
+        count = pixel & RLE_MAX_RUN;
+        if (!count)
+            return count;
+        if (pixel & RLE_COPY_FLAG) {
+            while(count--) {
+                *dest = *src++;
+                dest += z;
+            }
+        }
+        else {
+            while (count--) {
+                *dest = *src;
+                dest += z;
+            }
+            ++src;
+        }
+    }
+    return 0;
+}
+
+
+int
+ImagingSgiRleDecode(Imaging im, ImagingCodecState state,
+		    UINT8* buf, int bytes)
+{
+    UINT8 *ptr;
+    SGISTATE *c;
+
+    /* Get all data from File descriptor */    
+    c = (SGISTATE*)state->context;
+    _imaging_seek_pyFd(state->fd, 0L, SEEK_END);
+    c->bufsize = _imaging_tell_pyFd(state->fd);
+    c->bufsize -= SGI_HEADER_SIZE;
+    ptr = malloc(sizeof(UINT8) * c->bufsize);
+    _imaging_seek_pyFd(state->fd, SGI_HEADER_SIZE, SEEK_SET);
+    _imaging_read_pyFd(state->fd, (char*)ptr, c->bufsize);
+
+
+    /* decoder initialization */
+    state->count = 0;
+    state->y = 0;
+    if (state->ystep < 0)
+        state->y = im->ysize - 1;
+    else
+        state->ystep = 1;
+
+    /* Allocate memory for RLE tables and rows */
+    free(state->buffer);
+    state->buffer = malloc(sizeof(UINT8) * 2 * im->xsize * im->bands);
```

Comment:
Should probably just calloc this one as well. `calloc(im->xsize, im->bands * 2 )`

* sizeof(uint8) should be 1